libcxx/include/flat_set

```diff
@@ -0,0 +1,913 @@
+  flat_set(sorted_unique_t __s, initializer_list<value_type> __il, const _Alloc& __a)
+    : flat_set(__s, __il.begin(), __il.end(), __a) { }
+
+  flat_set& operator=(initializer_list<value_type> __il) {
+    clear();
+    insert(__il.begin(), __il.end());
+    return *this;
+  }
+
+  flat_set& operator=(const flat_set& __fs) {
+    auto __guard = std::__make_exception_guard([&] { __restore_invariant(); });
+    __compare_ = __fs.__compare_;
+    __c_ = __fs.__c_;
+    __guard.__complete();
+    return *this;
+  }
+
+  flat_set& operator=(flat_set&& __fs)
+    noexcept(is_nothrow_assignable_v<_Compare&, _Compare&> && is_nothrow_assignable_v<_KeyContainer, _KeyContainer&&>)
+  {
+    auto __guard = std::__make_exception_guard([&] { __restore_invariant(); });
+    __compare_ = __fs.__compare_;
+    __c_ = std::move(__fs.__c_);
+    __guard.__complete();
+    return *this;
+  }
+
+  // iterators
+
+  iterator begin() noexcept { return __c_.begin(); }
+  const_iterator begin() const noexcept { return __c_.begin(); }
+  iterator end() noexcept { return __c_.end(); }
+  const_iterator end() const noexcept { return __c_.end(); }
+
+  reverse_iterator rbegin() noexcept { return reverse_iterator(end()); }
+  const_reverse_iterator rbegin() const noexcept { return const_reverse_iterator(end()); }
+  reverse_iterator rend() noexcept { return reverse_iterator(begin()); }
+  const_reverse_iterator rend() const noexcept { return const_reverse_iterator(begin()); }
+
+  const_iterator cbegin() const noexcept { return begin(); }
+  const_iterator cend() const noexcept { return end(); }
+  const_reverse_iterator crbegin() const noexcept { return rbegin(); }
+  const_reverse_iterator crend() const noexcept { return rend(); }
+
+  // capacity
+
+  [[nodiscard]] bool empty() const noexcept { return __c_.empty(); }
+  size_type size() const noexcept { return __c_.size(); }
+  size_type max_size() const noexcept { return __c_.max_size(); }
+
+  // [flat.set.modifiers]
+
+  template <class... _Args>
+  pair<iterator, bool> emplace(_Args&&... __args) {
+    auto __key = key_type(std::forward<_Args>(__args)...);
+    auto __it = lower_bound(__key);
+    if (__it == end() || __compare_(__key, *__it)) {
+      auto __guard = std::__make_exception_guard([&] { __restore_invariant(); });
+      __it = __c_.emplace(__it, std::move(__key));
+      __guard.__complete();
+      return {__it, true};
+    } else {
+      return {__it, false};
+    }
+  }
+
+  template <class... _Args>
+  iterator emplace_hint(const_iterator, _Args&&... __args) {
+    return emplace(std::forward<_Args>(__args)...).first;
+  }
+
+  pair<iterator, bool> insert(const value_type& __x) { return emplace(__x); }
+  pair<iterator, bool> insert(value_type&& __x) { return emplace(std::move(__x)); }
+
+  template <class _K2>
+    requires __is_transparent<_Compare, _K2>::value &&
+             is_constructible_v<value_type, _K2>
+  pair<iterator, bool> insert(_K2&& __x) {
```

Comment:
We have a precondition that `find(value_type(__x)) == find(__x)` -- we could consider adding this in debug mode.

Reply:
Done; but notice that libc++ does not bother with that precondition on either [`set::find(K&&)`](https://github.com/llvm/llvm-project/blob/4f15267/libcxx/include/set#L847-L856) or [`unordered_set::find(K&&)`](https://github.com/llvm/llvm-project/blob/4f15267/libcxx/include/unordered_set#L1424-L1435).